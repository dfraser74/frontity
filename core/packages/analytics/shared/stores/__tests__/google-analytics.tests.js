import { types, unprotect } from 'mobx-state-tree';
// import GoogleAnalytics from '../google-analytics';
import Analytics from '..';

const itemSingle = {
  type: 'post',
  id: 60,
  entity: {
    headMeta: { title: 'The Beauties of Gullfoss' },
    link: 'https://demo.frontity.test/the-beauties-of-gullfoss/',
  },
};

const itemList = {
  type: 'category',
  id: 7,
  page: 2,
  entity: {
    headMeta: { title: 'Photography' },
    link: 'https://demo.frontity.test/wp-cat/photography/',
    pagedLink: page =>
      `https://demo.frontity.test/wp-cat/photography/page/${page}`,
  },
};

const Stores = types.model('Stores').props({
  connection: types.optional(types.frozen, {
    selectedItem: itemSingle,
    selectedContext: {
      options: {
        bar: 'single',
      },
    },
  }),
  analytics: types.optional(Analytics, {}),
});

let stores;
beforeEach(() => {
  stores = Stores.create({});
  unprotect(stores);
});

describe('Analytics > GoogleAnalytics', () => {
  test('init', async () => {
    // The Google Analytics snippet inserts its script before the first one,
    // so there must be an script in the DOM mock.
    const firstScript = window.document.createElement('script');
    window.document.body.appendChild(firstScript);

    window.ga = jest.fn();
    const gaTrackingIds = ['UA-123456', 'UA-778899'];
    await stores.analytics.googleAnalytics.init(gaTrackingIds);
    expect(stores.analytics.googleAnalytics).toMatchSnapshot();
    expect(window.ga).toHaveBeenCalledTimes(4);
    expect(window.ga.mock.calls).toMatchSnapshot();
  });

  test('sendPageView', () => {
    window.ga = jest.fn();
    stores.analytics.googleAnalytics.trackerNames = ['test1', 'test2'];
    stores.analytics.googleAnalytics.sendPageView();
    expect(window.ga).toHaveBeenCalledTimes(2);

    stores.connection = { selectedItem: itemList };
    stores.analytics.googleAnalytics.sendPageView();
    expect(window.ga).toHaveBeenCalledTimes(4);

    expect(window.ga.mock.calls).toMatchSnapshot();
  });

  test('sendEvent', () => {
    const event = {
      category: 'test/category',
      action: 'test/action',
      label: 'test/label',
    };

    window.ga = jest.fn();
    stores.analytics.googleAnalytics.trackerNames = ['test1', 'test2'];
    stores.analytics.googleAnalytics.sendEvent(event);
    expect(window.ga).toHaveBeenCalledTimes(2);
    expect(window.ga.mock.calls).toMatchSnapshot();
  });
});
