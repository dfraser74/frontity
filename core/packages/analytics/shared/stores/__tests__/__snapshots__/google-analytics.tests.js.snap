// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics > GoogleAnalytics init 1`] = `
Object {
  "ampTriggers": Object {},
  "ampVars": Object {},
}
`;

exports[`Analytics > GoogleAnalytics init 2`] = `
Array [
  Array [
    "create",
    "UA-123456",
    "auto",
    "clientTracker0",
  ],
  Array [
    "create",
    "UA-778899",
    "auto",
    "clientTracker1",
  ],
  Array [
    "clientTracker0.send",
    Object {
      "dimension1": "dim1_post60",
      "dimension2": "dim2_post60",
      "hitType": "pageview",
      "location": "https://demo.frontity.test/the-beauties-of-gullfoss/",
      "title": "The Beauties of Gullfoss",
    },
  ],
  Array [
    "clientTracker1.send",
    Object {
      "dimension1": "dim1_post60",
      "dimension2": "dim2_post60",
      "hitType": "pageview",
      "location": "https://demo.frontity.test/the-beauties-of-gullfoss/",
      "title": "The Beauties of Gullfoss",
    },
  ],
]
`;

exports[`Analytics > GoogleAnalytics sendEvent 1`] = `
Array [
  Array [
    "test1.send",
    Object {
      "eventAction": "PWA - test/action",
      "eventCategory": "PWA - test/category",
      "eventLabel": "test/label type: post context: single",
      "hitType": "event",
    },
  ],
  Array [
    "test2.send",
    Object {
      "eventAction": "PWA - test/action",
      "eventCategory": "PWA - test/category",
      "eventLabel": "test/label type: post context: single",
      "hitType": "event",
    },
  ],
]
`;

exports[`Analytics > GoogleAnalytics sendPageView 1`] = `
Array [
  Array [
    "test1.send",
    Object {
      "dimension1": "dim1_post60",
      "dimension2": "dim2_post60",
      "hitType": "pageview",
      "location": "https://demo.frontity.test/the-beauties-of-gullfoss/",
      "title": "The Beauties of Gullfoss",
    },
  ],
  Array [
    "test2.send",
    Object {
      "dimension1": "dim1_post60",
      "dimension2": "dim2_post60",
      "hitType": "pageview",
      "location": "https://demo.frontity.test/the-beauties-of-gullfoss/",
      "title": "The Beauties of Gullfoss",
    },
  ],
  Array [
    "test1.send",
    Object {
      "dimension1": "dim1_cat7",
      "dimension2": "dim2_cat7",
      "hitType": "pageview",
      "location": "https://demo.frontity.test/wp-cat/photography/page/2",
      "title": "Photography",
    },
  ],
  Array [
    "test2.send",
    Object {
      "dimension1": "dim1_cat7",
      "dimension2": "dim2_cat7",
      "hitType": "pageview",
      "location": "https://demo.frontity.test/wp-cat/photography/page/2",
      "title": "Photography",
    },
  ],
]
`;

exports[`Analytics > GoogleAnalytics trackingIds 1`] = `
Array [
  "UA-12345678-1",
  "UA-12345678-2",
]
`;

exports[`Analytics > GoogleAnalytics trackingIds 2`] = `
Array [
  "UA-12345678-1",
  "UA-12345678-2",
]
`;

exports[`Analytics > GoogleAnalytics trackingOptions 1`] = `
Array [
  Object {
    "sendEvents": true,
    "sendPageViews": true,
  },
  Object {
    "sendEvents": true,
    "sendPageViews": true,
  },
]
`;

exports[`Analytics > GoogleAnalytics trackingOptions 2`] = `
Array [
  Object {
    "sendEvents": true,
    "sendPageViews": true,
  },
  Object {
    "sendEvents": true,
    "sendPageViews": false,
  },
]
`;
